workers/docker/docker.go

```diff
@@ -0,0 +1,132 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"github.com/gaia-pipeline/gaia"
+	"github.com/gaia-pipeline/gaia/security"
+)
+
+// Worker represents the data structure of a docker worker.
+type Worker struct {
+	Host          string `json:"host"`
+	WorkerID      string `json:"worker_id"`
+	ContainerID   string `json:"container_id"`
+	PipelineRunID string `json:"pipeline_run_id"`
+}
+
+// NewDockerWorker initiates a new worker instance.
+func NewDockerWorker(host string, pipelineRunID string) *Worker {
+	return &Worker{Host: host, PipelineRunID: pipelineRunID}
+}
+
+// SetupDockerWorker starts a Gaia worker inside a docker container, automatically
+// connects it with this Gaia instance and sets a unique tag.
+func (w *Worker) SetupDockerWorker(workerImage string, workerSecret string) error {
+	// Generate a unique id for this worker
+	w.WorkerID = security.GenerateRandomUUIDV5()
+
+	// Setup docker client
+	ctx := context.Background()
+	cli, err := client.NewClientWithOpts(client.WithHost(w.Host))
+	if err != nil {
+		gaia.Cfg.Logger.Error("failed to setup docker client", "error", err)
+		return err
+	}
+	cli.NegotiateAPIVersion(ctx)
+
+	// Define small helper function which creates the docker container
+	createContainer := func() (container.ContainerCreateCreatedBody, error) {
+		return cli.ContainerCreate(ctx, &container.Config{
+			Image: workerImage,
+			Env: []string{
+				"GAIA_WORKER_HOST_URL=" + gaia.Cfg.DockerWorkerHostURL,
+				"GAIA_MODE=worker",
+				"GAIA_WORKER_GRPC_HOST_URL=" + gaia.Cfg.DockerWorkerGRPCHostURL,
+				"GAIA_WORKER_TAGS=" + fmt.Sprintf("%s,dockerworker", w.WorkerID),
+				"GAIA_WORKER_SECRET=" + workerSecret,
+			},
+		}, &container.HostConfig{}, nil, "")
+	}
+
+	// Create container
+	resp, err := createContainer()
+	if err != nil {
+		gaia.Cfg.Logger.Error("failed to create worker container", "error", err)
+		gaia.Cfg.Logger.Info("try to pull docker image and then try it again...")
+
+		// Pull worker image
+		_, err = cli.ImagePull(ctx, workerImage, types.ImagePullOptions{})
```

Comment:
It's worth mentioning that right now this only supports public docker registry and no private registries. I guess we can add that functionality later on?

Reply:
Good catch! Yes, exactly. I wouldn't prioritize this yet since we can easily add this at a later point in time.